File: sdk-libs/client/src/prover/merge.rs
use num_bigint::BigUint;
use solana_address::Address;
use zolana_event::MessageData;
use zolana_hasher::hash_chain::create_hash_chain_4_from_slice;
use zolana_interface::{
    error::ShieldedPoolError,
    instruction::{
        instruction_data::{
            merge_ring::MergeRingIxData,
            merge_transact::{MergeExternalDataHash, MergeProof, MergeTransactIxData},
        },
        tag::{MERGE_TRANSACT, RING_MERGE_TRANSACT},
    },
    state::cache::CACHE_CAPACITY,
    tree_slot::{tree_id_field, tree_slots_hash_chain},
};
use zolana_keypair::{Curve, NullifierKey};
use zolana_transaction::{
    instructions::merge::{
        merge_dummy_nullifier, merge_output_blinding, merge_private_tx_blinding, MergeProofInputs,
        MERGE_SUPPORTED_INPUT_COUNTS,
    },
    utxo::program_id_proof_input_hash,
};

use crate::{
    error::ClientError,
    prover::{
        field::{be, right_align, right_align_slice},
        transact::{
            assembly::{assemble_inputs, assemble_outputs, private_tx_hash, OwnerMode},
            witness::{attach_input_proofs, SpendProof},
        },
        MergeInputs, TreeSlotFields,
    },
    rpc::NonInclusionProof,
};

pub struct MergeProver {
    pub transaction: MergeProofInputs,
    pub nullifier_key: NullifierKey,
    pub proofs: Vec<SpendProof>,
    pub dummy_nullifier_proofs: Vec<NonInclusionProof>,
    pub cache: Option<MergeCacheTarget>,
}

/// The cache slot a merge writes its output to.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MergeCacheTarget {
    pub address: Address,
    pub slot: u8,
}

#[derive(Debug, Clone)]
pub struct MergeProofResult {
    pub inputs: MergeInputs,
    pub public_input_hash: [u8; 32],
    pub nullifiers: Vec<[u8; 32]>,
    /// Root cache indexes shared by all input slots.
    pub utxo_tree_root_index: u16,
    pub nullifier_tree_root_index: u16,
    pub output_hash: [u8; 32],
    pub private_tx_hash: [u8; 32],
    /// Recomputed on-chain from the instruction; surfaced so the caller need not
    /// re-derive it.
    pub external_data_hash: [u8; 32],
    pub expiry_unix_ts: u64,
    /// True when the owner is a Solana (ed25519) signer, so `merge_transact` derives
    /// `signing_pk_field` from the registry account owner instead of `owner_p256`.
    pub eddsa_owner: bool,
    pub cache_slot: Option<u8>,
    pub ring_program_id: Option<Address>,
    pub output_ring_data_hash: [u8; 32],
    pub tx_viewing_pk: [u8; 33],
    pub salt: [u8; 16],
    pub output_data: MessageData,
}

impl MergeProofResult {
    /// Assemble the `merge_transact` instruction data from this proof result and
    /// the proof (`ProofCompressed::to_merge_proof`). The caller passes the
    /// result to the `MergeTransact` builder with the tree / protocol_config /
    /// user_record accounts.
    pub fn instruction_data(&self, proof: MergeProof) -> MergeTransactIxData {
        MergeTransactIxData {
            expiry_unix_ts: self.expiry_unix_ts,
            proof,
            output_utxo_hash: self.output_hash,
            nullifiers: self.nullifiers.clone(),
            utxo_tree_root_index: self.utxo_tree_root_index,
            nullifier_tree_root_index: self.nullifier_tree_root_index,
            private_tx_hash: self.private_tx_hash,
            eddsa_owner: self.eddsa_owner,
            cache_slot: self.cache_slot,
        }
    }

    /// Assemble the `merge_ring` instruction data: the same `merge_transact`
    /// body wrapped in a [`MergeRingIxData`] with the output `ring_data_hash`
    /// the ring program selected. The caller passes the result to the
    /// `MergeRing` builder with the tree / ring_config accounts.
    pub fn ring_instruction_data(&self, proof: MergeProof) -> MergeRingIxData {
        MergeRingIxData {
            output_ring_data_hash: self.output_ring_data_hash,
            merge: self.instruction_data(proof),
        }
    }
}

impl MergeProver {
    pub fn build(self) -> Result<MergeProofResult, ClientError> {
        let tx = &self.transaction;
        let n_inputs = tx.input_utxos.len();
        if !MERGE_SUPPORTED_INPUT_COUNTS.contains(&n_inputs) {
            return Err(ClientError::UnsupportedShape {
                n_in: n_inputs,
                n_out: 1,
            });
        }
        let first = tx
            .input_utxos
            .first()
            .filter(|input| !input.is_dummy())
            .ok_or(ClientError::NoInputs)?;
        let first_nullifier = first.nullifier;
        let nullifier_pubkey = self.nullifier_key.pubkey()?;
        let mut total = 0u64;
        tx.input_utxo_hashes()?;
        for (index, input) in tx.input_utxos.iter().enumerate() {
            if input.is_dummy() {
                let slot = u8::try_from(index).map_err(|_| ClientError::TooManyInputs {
                    got: n_inputs,
                    max: usize::from(u8::MAX),
                })?;
                if input.nullifier
                    != merge_dummy_nullifier(&self.nullifier_key, &first_nullifier, slot)?
                {
                    return Err(ClientError::InputNullifierMismatch { index });
                }
                continue;
            }
            if input.utxo.owner != tx.signing_pubkey {
                return Err(ClientError::MergeSigningKeyMismatch);
            }
            if input.nullifier_pubkey != nullifier_pubkey {
                return Err(ClientError::MergeNullifierKeyMismatch);
            }
            if input.utxo.ring_program_id != tx.ring_program_id {
                return Err(ClientError::InputRingProgramMismatch { index });
            }
            if input.utxo.asset != first.utxo.asset {
                return Err(ClientError::MergeInputAssetMismatch { index });
            }
            if input.nullifier
                != self
                    .nullifier_key
                    .nullifier(&input.utxo_hash, &input.utxo.blinding)?
            {
                return Err(ClientError::InputNullifierMismatch { index });
            }
            total = total
                .checked_add(input.utxo.amount)
                .ok_or(ClientError::SelectedBalanceOverflow)?;
        }
        let output = &tx.output_utxo;
        if output.is_dummy()
            || output.asset != first.utxo.asset
            || output.amount != total
            || output.ring_program_id != tx.ring_program_id
            || output.data_hash.is_some()
            || (tx.ring_program_id.is_none() && output.ring_data_hash.is_some())
            || !output.owner_address.is_some_and(|owner| {
                owner.signing_pubkey == tx.signing_pubkey
                    && owner.nullifier_pubkey == nullifier_pubkey
            })
        {
            return Err(ClientError::MergeOutputMismatch);
        }
        if output.blinding != merge_output_blinding(&self.nullifier_key, &first_nullifier)? {
            return Err(ClientError::OutputBlindingMismatch { index: 0 });
        }
        let MergeProofInputs {
            input_utxos,
            output_utxo,
            expiry_unix_ts,
            signing_pubkey,
            output_tree_id,
            ring_program_id,
            tx_viewing_pk,
            salt,
            output_data,
        } = self.transaction;
        let inputs = attach_input_proofs(input_utxos, &self.proofs, &self.dummy_nullifier_proofs)?;
        let assembled_inputs = assemble_inputs(&inputs, &OwnerMode::Merge)?;
        let input_tree_context = assembled_inputs.single_tree_context()?;
        let assembled_outputs =
            assemble_outputs(std::slice::from_ref(&output_utxo), output_tree_id)?;
        let output_hash = *assembled_outputs
            .output_hashes
            .first()
            .ok_or(ClientError::MissingOutput)?;
        let cache_slot = match self.cache.as_ref() {
            Some(target) if usize::from(target.slot) < CACHE_CAPACITY => Some(target.slot),
            Some(_) => return Err(ShieldedPoolError::InvalidCacheSlot.into()),
            None => None,
        };
        let external_data_hash = MergeExternalDataHash {
            spp_instruction_discriminator: if ring_program_id.is_some() {
                RING_MERGE_TRANSACT
            } else {
                MERGE_TRANSACT
            },
            expiry_unix_ts,
            output_utxo_hash: &output_hash,
            cache: self
                .cache
                .as_ref()
                .map(|target| target.address.as_array())
                .zip(cache_slot),
        }
        .hash()?;
        let private_tx_blinding = merge_private_tx_blinding(&self.nullifier_key, &first_nullifier)?;
        let private_tx = private_tx_hash(
            &assembled_inputs,
            &assembled_outputs,
            &external_data_hash,
            &private_tx_blinding,
        )?;
        let user_signing_pk_hash = signing_pubkey.owner_proof_input_hash()?;
        let mut elements = vec![
            create_hash_chain_4_from_slice(&assembled_inputs.nullifiers)?,
            output_hash,
            tree_slots_hash_chain(&assembled_inputs.tree_slots)?,
            tree_id_field(output_tree_id),
            private_tx,
            external_data_hash,
            right_align(&[1u8]),
        ];
        let output_ring_data_hash = output_utxo.ring_data_hash.unwrap_or_default();
        let ring_hash = program_id_proof_input_hash(&ring_program_id)?;
        if ring_program_id.is_some() {
            elements.extend([output_ring_data_hash, ring_hash]);
        } else {
            // Bind both halves of the UTXO owner to the registry, so another
            // nullifier key cannot manufacture a merge for this signing identity.
            elements.extend([user_signing_pk_hash, nullifier_pubkey]);
        }
        let public_input_hash = create_hash_chain_4_from_slice(&elements)?;
        let eddsa_owner = match signing_pubkey.curve()? {
            Curve::Ed25519 | Curve::Pda => true,
            Curve::P256 => false,
        };
        let user_nullifier_secret = right_align_slice(&*self.nullifier_key.secret())?;
        let output = assembled_outputs
            .outputs
            .into_iter()
            .next()
            .ok_or(ClientError::MissingOutput)?;
        let inputs = MergeInputs {
            inputs: assembled_inputs.inputs,
            output,
            tree_slots: TreeSlotFields::encode_all(&assembled_inputs.tree_slots),
            output_tree_id: BigUint::from(output_tree_id),
            owner_pk_hash: be(&user_signing_pk_hash),
            user_nullifier_pk: be(&nullifier_pubkey),
            user_nullifier_secret: be(&user_nullifier_secret),
            external_data_hash: be(&external_data_hash),
            private_tx_hash: be(&private_tx),
            allow_dummy_inputs: BigUint::from(1u8),
            public_input_hash: be(&public_input_hash),
            output_ring_data_hash: be(&output_ring_data_hash),
            ring_program_id: be(&ring_hash),
        };
        Ok(MergeProofResult {
            inputs,
            public_input_hash,
            nullifiers: assembled_inputs.nullifiers,
            utxo_tree_root_index: input_tree_context.utxo_tree_root_index,
            nullifier_tree_root_index: input_tree_context.nullifier_tree_root_index,
            output_hash,
            private_tx_hash: private_tx,
            external_data_hash,
            expiry_unix_ts,
            eddsa_owner,
            cache_slot,
            ring_program_id,
            output_ring_data_hash,
            tx_viewing_pk,
            salt,
            output_data,
        })
    }
}

File: sdk-libs/ts/src/client/prover/merge.ts
import type {
  Address,
  Bytes32,
  MergeTransactInstructionData,
  RequestContext,
} from "../../interface/types.js";
import { mergeExternalDataHash } from "../../interface/codecs/index.js";
import { InstructionTag } from "../../interface/program.js";
import { treeAddress } from "../../interface/pda/index.js";
import { inputTreeSlots, treeIdField, type TreeSlot } from "../../interface/tree-slot.js";
import { MERGE_INPUTS, PreparedMerge } from "../../transaction/instructions/builders.js";

import type { ProofReader, IndexedProofInputs, PreparedMergeInputs } from "../ports.js";
import { ClientError, fromClientCause } from "../error.js";
import {
  bigintToBytes,
  bytesField,
  bytesToBigInt,
  checkedBytes,
  field,
  hashChain4,
  poseidon,
} from "../internal.js";
import type { NonInclusionProof, SpendProof } from "../rpc.js";
import {
  createDummyTransferInput,
  createOutput,
  createRealInput,
  prepareInput,
  resolvedPublicInputHash,
  treeSlotFields,
  validateSpendProof,
} from "./assembly.js";
import type { Field, MergeInputs, TransferInput } from "./types.js";

export interface MergeAssembly {
  readonly proverInputs: MergeInputs;
  readonly expiryUnixTs: bigint;
  readonly outputHash: Bytes32;
  readonly nullifiers: readonly Bytes32[];
  readonly utxoTreeRootIndex: number;
  readonly nullifierTreeRootIndex: number;
  readonly privateTxHash: Bytes32;
  readonly publicInputHash: Bytes32;
  /// Recomputed on-chain from the instruction; surfaced so the caller need not
  /// re-derive it.
  readonly externalDataHash: Bytes32;
  readonly eddsaOwner: boolean;
  instructionData(proof: MergeTransactInstructionData["proof"]): MergeTransactInstructionData;
}

export async function assembleMerge(
  prepared: PreparedMerge,
  indexer: Pick<ProofReader, "getInputMerkleProofs" | "getNonInclusionProofs">,
  tree: Address,
  context?: RequestContext,
): Promise<MergeAssembly> {
  try {
    validatePreparedMerge(prepared);
    const dummyNullifiers = prepared.dummyNullifiers();
    const [proofs, dummyResponse] = await Promise.all([
      indexer.getInputMerkleProofs(prepared.inputUtxoHashes(), undefined, context),
      dummyNullifiers.length === 0
        ? Promise.resolve(undefined)
        : indexer.getNonInclusionProofs(tree, dummyNullifiers, undefined, context),
    ]);
    return assembleMergeUnchecked(prepared, proofs, dummyResponse?.proofs ?? [], tree);
  } catch (cause) {
    throw fromClientCause(cause);
  }
}

export function assembleMergeWithProofs(
  prepared: PreparedMerge,
  proofs: readonly SpendProof[],
  tree: Address,
  dummyNullifierProofs: readonly NonInclusionProof[] = [],
): MergeAssembly {
  try {
    return assembleMergeUnchecked(prepared, proofs, dummyNullifierProofs, tree);
  } catch (cause) {
    throw fromClientCause(cause);
  }
}

/** The one tree slot a merge opens against, and the root positions its instruction references. */
interface MergeInputTree {
  readonly slot: TreeSlot;
  readonly utxoRootIndex: number;
  readonly nullifierRootIndex: number;
}

function assembleMergeUnchecked(
  prepared: PreparedMerge,
  proofs: readonly SpendProof[],
  dummyNullifierProofs: readonly NonInclusionProof[],
  tree: Address,
): MergeAssembly {
  validateMergeTree(prepared, tree);
  const realInputs = prepared.inputs.filter((input) => !input.isDummy());
  if (proofs.length !== realInputs.length) {
    throw new ClientError("CLIENT_INCOMPLETE_INPUT_PROOFS", {
      details: { expected: realInputs.length, state: proofs.length, nullifier: proofs.length },
    });
  }
  if (realInputs.length === 0) throw new ClientError("CLIENT_NO_INPUTS");
  const dummyNullifiers = prepared.dummyNullifiers();
  if (dummyNullifierProofs.length !== dummyNullifiers.length) {
    throw new ClientError("CLIENT_INCOMPLETE_INPUT_PROOFS", {
      details: {
        expected: dummyNullifiers.length,
        state: 0,
        nullifier: dummyNullifierProofs.length,
      },
    });
  }
  const inputs: TransferInput[] = [];
  let inputTree: MergeInputTree | undefined;
  let proofIndex = 0;
  let dummyIndex = 0;
  for (const [index, input] of prepared.inputs.entries()) {
    if (input.isDummy()) {
      if (inputTree === undefined) throw new ClientError("CLIENT_NO_INPUTS");
      const nullifier = dummyNullifiers[dummyIndex];
      const proof = dummyNullifierProofs[dummyIndex++];
      if (!nullifier || !proof) {
        throw new ClientError("CLIENT_MISSING_INPUT_MERKLE_PROOF", {
          details: { index: dummyIndex - 1 },
        });
      }
      if (!equal(proof.leaf, nullifier)) {
        throw new ClientError("CLIENT_NULLIFIER_PROOF_LEAF_MISMATCH", {
          details: { index: dummyIndex - 1 },
        });
      }
      if (proof.merkleContext.tree !== tree) {
        throw new ClientError("CLIENT_MERGE_TREE_MISMATCH", {
          details: { proofTree: proof.merkleContext.tree, submitTree: tree },
        });
      }
      checkNullifierRoot(inputTree, proof, index);
      const converted = createDummyTransferInput(input, proof, nullifier);
      inputs.push(converted);
      continue;
    }
    const proof = proofs[proofIndex];
    if (!proof) {
      throw new ClientError("CLIENT_MISSING_INPUT_MERKLE_PROOF", {
        details: { index: proofIndex },
      });
    }
    validateSpendProof(input, proof, proofIndex);
    if (proof.state.merkleContext.tree !== tree) {
      throw new ClientError("CLIENT_MERGE_TREE_MISMATCH", {
        details: {
          proofTree: proof.state.merkleContext.tree,
          submitTree: tree,
        },
      });
    }
    if (proof.nullifier.merkleContext.tree !== tree) {
      throw new ClientError("CLIENT_MERGE_TREE_MISMATCH", {
        details: {
          proofTree: proof.nullifier.merkleContext.tree,
          submitTree: tree,
        },
      });
    }
    if (inputTree === undefined) {
      inputTree = Object.freeze({
        slot: Object.freeze({
          id: prepared.inputTreeId,
          utxoRoot: new Uint8Array(proof.state.root) as Bytes32,
          nullifierRoot: new Uint8Array(proof.nullifier.root) as Bytes32,
        }),
        utxoRootIndex: proof.state.rootIndex,
        nullifierRootIndex: proof.nullifier.rootIndex,
      });
    } else {
      if (
        !equal(proof.state.root, inputTree.slot.utxoRoot) ||
        proof.state.rootIndex !== inputTree.utxoRootIndex
      ) {
        throw new ClientError("CLIENT_INPUT_TREE_ROOT_MISMATCH", { details: { index } });
      }
      checkNullifierRoot(inputTree, proof.nullifier, index);
    }
    // A P256 owner contributes the 0 sentinel: the merge circuit recomputes its
    // pk_field from the witnessed point and ignores the per-input value.
    const ownerPublicKeyHash =
      input.utxo.owner.signatureType() === "p256"
        ? 0n
        : bytesField(input.utxo.owner.ownerProofInputHash(), "merge owner public key");
    const converted = createRealInput(input, proof, ownerPublicKeyHash);
    inputs.push(converted);
    proofIndex++;
  }
  if (inputTree === undefined) throw new ClientError("CLIENT_NO_INPUTS");

  const local = prepareMerge(prepared, tree);
  const complete = local.finish(inputTree);
  return Object.freeze({
    ...complete,
    proverInputs: Object.freeze({
      ...local.inputs.payload,
      inputs: Object.freeze(inputs),
      treeSlots: Object.freeze(inputTreeSlots([inputTree.slot]).map(treeSlotFields)),
      publicInputHash: asField(bytesToBigInt(complete.publicInputHash)),
    }),
  });
}

interface PreparedMergeAssembly {
  readonly inputs: IndexedProofInputs & {
    readonly circuit: "merge";
    readonly payload: PreparedMergeInputs;
  };
  finish(tree: MergeInputTree): Omit<MergeAssembly, "proverInputs">;
}

export function prepareMerge(prepared: PreparedMerge, tree: Address): PreparedMergeAssembly {
  validateMergeTree(prepared, tree);
  const expiryUnixTs = prepared.expiryUnixTs;
  // 1. Reuse commitments within one call to keep mutable input bytes bound to the statement.
  const realCommitments = new Map(
    prepared.inputUtxoHashes().map(({ index, utxoHash }) => [index, utxoHash]),
  );
  const commitments = prepared.inputs.map((input, index) => {
    if (input.isDummy()) return null;
    const commitment = realCommitments.get(index);
    if (commitment === undefined) throw new ClientError("CLIENT_INVALID_MERGE");
    return commitment;
  });
  const dummyNullifiers = prepared.dummyNullifiers();
  let dummyIndex = 0;
  const inputs = prepared.inputs.map((input) => {
    const nullifier = input.isDummy() ? dummyNullifiers[dummyIndex++] : input.nullifier();
    if (nullifier === undefined) throw new ClientError("CLIENT_INVALID_MERGE");
    const owner =
      input.isDummy() || input.utxo.owner.signatureType() === "p256"
        ? 0n
        : bytesField(input.utxo.owner.ownerProofInputHash(), "merge owner public key");
    return prepareInput(input, { owner, treeSlot: 0, nullifier });
  });
  const inputHashes = commitments.map((commitment) =>
    commitment === null ? 0n : bytesToBigInt(commitment),
  );
  const nullifiers = inputs.map((input) =>
    checkedBytes(bigintToBytes(input.nullifier), 32, "nullifier"),
  );
  const output = createOutput(prepared.output, prepared.outputTreeId);
  if (prepared.output.isDummy()) throw new ClientError("CLIENT_INVALID_MERGE_OUTPUT");
  const outputHash = checkedBytes(prepared.outputHash(), 32, "merge output hash");
  const externalDataHash = mergeExternalDataHash({
    instructionTag:
      prepared.output.ringProgramId === undefined
        ? InstructionTag.mergeTransact
        : InstructionTag.ringMergeTransact,
    expiryUnixTs,
    outputUtxoHash: outputHash,
  });
  // Merge has no blinding seed: the owner's nullifier secret takes its place
  // in the private transaction blinding, so a reader holding the secret
  // recovers the output without any disclosed value.
  const firstNullifier = nullifiers[0];
  if (firstNullifier === undefined) throw new ClientError("CLIENT_NO_INPUTS");
  const privateTxBlinding = prepared.privateTxBlinding();
  const privateTxHash = bigintToBytes(
    poseidon([
      hashChain4(inputHashes),
      bytesToBigInt(outputHash),
      hashChain4(Array.from({ length: MERGE_INPUTS }, () => 0n)),
      bytesToBigInt(externalDataHash),
      bytesField(privateTxBlinding, "merge private tx blinding"),
    ]),
  ) as Bytes32;
  const eddsaOwner = prepared.signingPublicKey.signatureType() === "ed25519";
  const ownerPublicKeyHash = bytesField(
    prepared.signingPublicKey.ownerProofInputHash(),
    "merge owner public key",
  );
  const outputTreeIdField = bytesToBigInt(treeIdField(prepared.outputTreeId));
  const publicInputs = [
    hashChain4(nullifiers.map(bytesToBigInt)),
    bytesToBigInt(outputHash),
    outputTreeIdField,
    bytesToBigInt(privateTxHash),
    bytesToBigInt(externalDataHash),
    1n,
    ...(prepared.output.ringProgramId === undefined
      ? [ownerPublicKeyHash]
      : [BigInt(output.circuit.ringDataHash), BigInt(output.circuit.ringProgramId)]),
  ].map(asField);
  const payload: PreparedMergeInputs = Object.freeze({
    inputs: Object.freeze(inputs),
    output,
    outputTreeId: asField(outputTreeIdField),
    ownerPublicKeyHash: asField(ownerPublicKeyHash),
    userNullifierPublicKey: asField(
      bytesField(prepared.nullifierPublicKey, "merge nullifier public key"),
    ),
    externalDataHash: asField(bytesToBigInt(externalDataHash)),
    privateTxHash: asField(bytesToBigInt(privateTxHash)),
    allowDummyInputs: asField(1n),
    outputRingDataHash: output.circuit.ringDataHash,
    ringProgramId: output.circuit.ringProgramId,
  });
  return Object.freeze({
    inputs: Object.freeze({
      circuit: "merge",
      payload,
      trees: Object.freeze([{ tree, id: prepared.inputTreeId }]),
      lookups: Object.freeze(
        commitments.map((commitment) => Object.freeze({ treeSlot: 0, commitment })),
      ),
      publicInputs: Object.freeze(publicInputs),
    }),
    finish(inputTree: MergeInputTree): Omit<MergeAssembly, "proverInputs"> {
      const publicInputHash = checkedBytes(
        bigintToBytes(resolvedPublicInputHash(publicInputs, inputTreeSlots([inputTree.slot]))),
        32,
        "public input hash",
      );
      const utxoTreeRootIndex = inputTree.utxoRootIndex;
      const nullifierTreeRootIndex = inputTree.nullifierRootIndex;
      const instructionData = (
        proof: MergeTransactInstructionData["proof"],
      ): MergeTransactInstructionData =>
        Object.freeze({
          expiryUnixTs,
          proof: copyMergeProof(proof),
          outputUtxoHash: new Uint8Array(outputHash) as Bytes32,
          eddsaOwner,
          privateTxHash: new Uint8Array(privateTxHash) as Bytes32,
          nullifiers: Object.freeze(
            nullifiers.map((nullifier) => new Uint8Array(nullifier) as Bytes32),
          ),
          utxoTreeRootIndex,
          nullifierTreeRootIndex,
        });
      return Object.freeze({
        expiryUnixTs,
        // `Object.freeze` seals the assembly and the nullifier array but not the
        // buffers inside them, and those are the buffers `instructionData` copies
        // from on every call. Hand out copies of everything the closure reads so a
        // frozen assembly cannot be steered into emitting different instruction
        // data than the one it was proved with.
        outputHash: new Uint8Array(outputHash) as Bytes32,
        nullifiers: Object.freeze(
          nullifiers.map((nullifier) => new Uint8Array(nullifier) as Bytes32),
        ),
        utxoTreeRootIndex,
        nullifierTreeRootIndex,
        privateTxHash: new Uint8Array(privateTxHash) as Bytes32,
        publicInputHash,
        externalDataHash,
        eddsaOwner,
        instructionData,
      });
    },
  });
}

function validateMergeTree(prepared: PreparedMerge, tree: Address): void {
  validatePreparedMerge(prepared);
  // The submit tree must be the tree the inputs are hashed under, or the proof
  // and the instruction would name different trees.
  if (treeAddress(prepared.inputTreeId) !== tree) {
    throw new ClientError("CLIENT_MERGE_TREE_MISMATCH", {
      details: { proofTree: treeAddress(prepared.inputTreeId), submitTree: tree },
    });
  }
  // The merge instruction appends its output to the same tree it spends from,
  // so an output hashed under another tree would prove a commitment the
  // instruction's output tree rejects.
  if (
    prepared.output.ringProgramId === undefined &&
    prepared.outputTreeId !== prepared.inputTreeId
  ) {
    throw new ClientError("CLIENT_TREE_ID_MISMATCH", {
      details: { expected: prepared.inputTreeId, actual: prepared.outputTreeId },
    });
  }
}

function checkNullifierRoot(
  inputTree: MergeInputTree,
  proof: NonInclusionProof,
  index: number,
): void {
  if (
    !equal(proof.root, inputTree.slot.nullifierRoot) ||
    proof.rootIndex !== inputTree.nullifierRootIndex
  ) {
    throw new ClientError("CLIENT_NULLIFIER_ROOT_MISMATCH", { details: { index } });
  }
}

function validatePreparedMerge(prepared: PreparedMerge): void {
  if (!(prepared instanceof PreparedMerge)) throw new ClientError("CLIENT_INVALID_MERGE");
  if (prepared.inputs.length !== MERGE_INPUTS) {
    throw new ClientError("CLIENT_INVALID_MERGE_SHAPE", {
      details: { expected: MERGE_INPUTS, actual: prepared.inputs.length },
    });
  }
  let total = 0n;
  prepared.inputs.forEach((input) => {
    if (!input.isDummy()) {
      if (
        input.utxo.ringProgramId !== prepared.output.ringProgramId ||
        input.utxo.asset !== prepared.output.asset
      )
        throw new ClientError("CLIENT_INVALID_MERGE");
      total += input.utxo.amount;
    }
    if (!input.isDummy() && !equal(input.nullifierPublicKey, prepared.nullifierPublicKey)) {
      throw new ClientError("CLIENT_MERGE_NULLIFIER_KEY_MISMATCH");
    }
    if (
      !input.isDummy() &&
      !equal(input.utxo.owner.toBytes(), prepared.signingPublicKey.toBytes())
    ) {
      throw new ClientError("CLIENT_MERGE_SIGNING_KEY_MISMATCH");
    }
  });
  if (
    !equal(
      prepared.output.ownerAddress?.signingPublicKey.toBytes() ?? new Uint8Array(),
      prepared.signingPublicKey.toBytes(),
    )
  )
    throw new ClientError("CLIENT_INVALID_MERGE_OUTPUT");
  if (total !== prepared.output.amount || total > 0xffff_ffff_ffff_ffffn)
    throw new ClientError("CLIENT_INVALID_MERGE_OUTPUT");
}

function copyMergeProof(
  proof: MergeTransactInstructionData["proof"],
): MergeTransactInstructionData["proof"] {
  return Object.freeze({
    a: checkedBytes(proof.a, 32, "merge proof a"),
    b: checkedBytes(proof.b, 128, "merge proof b"),
    c: checkedBytes(proof.c, 32, "merge proof c"),
  });
}

function asField(value: bigint): Field {
  return field(value, "merge field") as Field;
}

function equal(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) return false;
  let difference = 0;
  for (let index = 0; index < left.length; index++) {
    difference |= (left[index] ?? 0) ^ (right[index] ?? 0);
  }
  return difference === 0;
}
